core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -473,6 +473,283 @@ class DAGSchedulerSuite
+      // Complete all the tasks for the current attempt of stage 0 successfully
+      completeNextShuffleMapSuccesfully(0, attempt, numShufflePartitions = 2)
+
+      if (attempt < Stage.MAX_CONSECUTIVE_FAILURES / 2) {
+        // Now we should have a new taskSet, for a new attempt of stage 1.
+        // Fail all these tasks with FetchFailure
+        completeNextStageWithFetchFailure(1, attempt, shuffleDepOne)
+      } else {
+        completeNextShuffleMapSuccesfully(1, attempt, numShufflePartitions = 1)
+
+        // Fail stage 2
+        completeNextStageWithFetchFailure(2, attempt - Stage.MAX_CONSECUTIVE_FAILURES / 2,
+          shuffleDepTwo)
+      }
+
+      // this will (potentially) trigger a resubmission of stage 0, since we've lost some of its
+      // map output, for the next iteration through the loop
+      scheduler.resubmitFailedStages()
+    }
+
+    completeNextShuffleMapSuccesfully(0, 4, numShufflePartitions = 2)
+    completeNextShuffleMapSuccesfully(1, 4, numShufflePartitions = 1)
+
+    // Succeed stage2 with a "42"
+    completeNextResultStageWithSuccess(2, Stage.MAX_CONSECUTIVE_FAILURES/2)
+
+    assert(results === Map(0 -> 42))
+    assertDataStructuresEmpty()
+  }
+
+  /**
+   * In this test we demonstrate that only consecutive failures trigger a stage abort. A stage may
+   * fail multiple times, succeed, then fail a few more times (because its run again by downstream
+   * dependencies). The total number of failed attempts for one stage will go over the limit,
+   * but that doesn't matter, since they have successes in the middle.
+   */
+  test("Non-consecutive stage failures don't trigger abort") {
+    setupStageAbortTest(sc)
+
+    val shuffleOneRdd = new MyRDD(sc, 2, Nil).cache()
+    val shuffleDepOne = new ShuffleDependency(shuffleOneRdd, null)
+    val shuffleTwoRdd = new MyRDD(sc, 2, List(shuffleDepOne)).cache()
+    val shuffleDepTwo = new ShuffleDependency(shuffleTwoRdd, null)
+    val finalRdd = new MyRDD(sc, 1, List(shuffleDepTwo))
+    submit(finalRdd, Array(0))
+
+    for (attempt <- 0 until Stage.MAX_CONSECUTIVE_FAILURES-1) {
+      // Make each task in stage 0 success
+      completeNextShuffleMapSuccesfully(0, attempt, numShufflePartitions = 2)
+
+      // Now we should have a new taskSet, for a new attempt of stage 1.
+      // Fail these tasks with FetchFailure
+      completeNextStageWithFetchFailure(1, attempt, shuffleDepOne)
+
+      scheduler.resubmitFailedStages()
+
+      // Confirm we have not yet aborted
+      assert(scheduler.runningStages.nonEmpty)
+      assert(!ended)
+    }
+
+    // Rerun stage 0 and 1
+    completeNextShuffleMapSuccesfully(0, 3, numShufflePartitions = 2)
+    completeNextShuffleMapSuccesfully(1, 3, numShufflePartitions = 1)
+
+    // Fail stage 2
+    completeNextStageWithFetchFailure(2, 0, shuffleDepTwo)
+
+    scheduler.resubmitFailedStages()
+    // Rerun stage 0
+    completeNextShuffleMapSuccesfully(0, 4, numShufflePartitions = 2)
```

Comment:
@andrewor14 I don't think there is an easy way to just keep re-running stage 1.  After you've completed it successfully, your only way to trigger it again is to have a downstream stage with a fetch failure.  We can't just submit another job which has the same dependency, b/c it will actually be a different "skipped" stage initially (which now that I think about it, diminishes the need for this consecutive failure check, but still seems good, especially if we get rid of skipped stages).  So thats why we have to fail stage 2.

We don't *have* to retry stage 0, but its really the natural way stage retry works when you have multiple consecutive shuffle map stages, you end up with a cascade of stage retries.  Stage 2 has a fetch failure when reading from node X, so it says, I've lost all my input from stage 1 that used to live on node X, so stage 1 had better re-run to regenerate that input.  But then when stage 1 is about to run, it says, wait a minute, I've lost *my* input from stage 0 that used to live on node X.  So it tells stage 0 to regenerate that input.

We could avoid this situation by carefully putting all the map output from stage 0 on some set of nodes, and generating the map output for stage 1 on at least one additional node, and making sure the fetch failure in stage 2 is on the node which has no shuffle map output for stage 1.  But I think that would both make the test setup a lot more complicated, and also avoid the more natural cascading failures that are typical with fetch failures.  (In fact, this test doesn't exactly *choose* to return stage 0, that is just what is scheduled so it has to be done.)

I agree this is complicated, but I'm not sure what to add.  Maybe a comment on why we have to rerun stage 0 after the failure of stage 2: because we've lost all the map output on the lost host, that includes the input for stage 1, which has to be regenerated by stage 0 ?